src/python/turicreate/toolkits/image_analysis/image_analysis.py

```diff
@@ -192,6 +192,171 @@ def resize(image, width, height, channels=None, decode=False, resample="nearest"
+    >>> resized_img = turicreate.image_analysis.resize_with_original_aspect_ratio(img,min_side_length=None,max_side_length=500, channels=1)
+
+    Resize an SArray of images
+
+    >>> url ='https://static.turi.com/datasets/images/nested'
+    >>> image_sframe = turicreate.image_analysis.load_images(url, "auto", with_path=False,
+    ...                                                    recursive=True)
+    >>> image_sarray = image_sframe["image"]
+    >>> resized_images = turicreate.image_analysis.resize(image_sarray, min_side_length=500,max_side_length=None, channels=1)
+    >>> resized_images = turicreate.image_analysis.resize(image_sarray, min_side_length=None,max_side_length=500, channels=1)
+    """
+
+    from ...data_structures.sarray import SArray as _SArray
+    import turicreate as _tc
+
+    if min_side_length is None and max_side_length is None:
+        raise ValueError("Cannot resize when neither `min_side_length` or `max_side_length` is not provided")
+
+    if (min_side_length is not None and min_side_length > 0) and (max_side_length is not None and max_side_length > 0):
+        raise ValueError("Cannot provide both parameters `min_side_length` and `max_side_length` as only one is required to maintain the aspect ratio")
+
+    if min_side_length is not None:
+        if min_side_length > 0:
+            side_length_type = "min_length"
+            side_length = min_side_length
+        else:
+            raise ValueError("Value of min_side_length parameter cannot be negetive")
+    elif max_side_length is not None:
+        if max_side_length > 0:
+            side_length_type = "max_length"
+            side_length = max_side_length
+        else:
+            raise ValueError("Value of max_side_length parameter cannot be negetive")
+    else:
+        raise ValueError("Need to provode either the `min_side_length` or `max_side_length` to resize while maintaining the aspect ratio")
+    
+    if type(image) is _Image:
+        new_image_shape = _get_new_image_dimensions_with_original_aspect_ratio(image, side_length_type, side_length)
+        return resize(image, new_image_shape[0], new_image_shape[1], channels=channels, decode=decode, resample=resample)
+    elif type(image) is _SArray:
+        new_image_shape_sarray = image.apply(
+            lambda x: _get_new_image_dimensions_with_original_aspect_ratio(x, side_length_type, side_length)
+        )
+        image_with_new_dims_sframe = _tc.SFrame({"image": image, "new_shape": new_image_shape_sarray})
+        return image_with_new_dims_sframe.apply(
+            lambda x: resize(x['image'], x['new_shape'][0], x['new_shape'][1], channels=channels, decode=decode, resample=resample)
+        )
+    else:
+        raise ValueError(
+            "Cannot call 'resize' on objects that are not either an Image or SArray of Images"
+        )
+
+
+def _get_new_image_dimensions_with_original_aspect_ratio(image, side_length_type, side_length):
```

Comment:
It would be cleaner to make this an inner function to `resize_with_original_aspect_ratio`.

Reply:
Will do.